src/test/java/org/opensearch/knn/jni/JNIServiceTests.java

```diff
@@ -700,19 +709,98 @@ public void testQueryIndex_faiss_valid() throws IOException {
                 assertNotEquals(0, pointer);
 
                 for (float[] query : testData.queries) {
-                    KNNQueryResult[] results = JNIService.queryIndex(pointer, query, k, FAISS_NAME, null);
+                    KNNQueryResult[] results = JNIService.queryIndex(pointer, query, k, FAISS_NAME, null, null);
                     assertEquals(k, results.length);
                 }
 
                 // Filter will result in no ids
                 for (float[] query : testData.queries) {
-                    KNNQueryResult[] results = JNIService.queryIndex(pointer, query, k, FAISS_NAME, new int[] { 0 });
+                    KNNQueryResult[] results = JNIService.queryIndex(pointer, query, k, FAISS_NAME, new int[] { 0 }, null);
                     assertEquals(0, results.length);
                 }
             }
         }
     }
 
+    public void testQueryIndex_faiss_parentIds() throws IOException {
+
+        int k = 100;
+
+        List<String> methods = ImmutableList.of(faissMethod);
+        List<SpaceType> spaces = ImmutableList.of(SpaceType.L2, SpaceType.INNER_PRODUCT);
+        int[] parentIds = toParentIdArray(testDataNested.indexData.docs);
+        Map<Integer, Integer> idToParentIdMap = toIdToParentIdMap(testDataNested.indexData.docs);
+        for (String method : methods) {
+            for (SpaceType spaceType : spaces) {
+                Path tmpFile = createTempFile();
+                JNIService.createIndex(
+                    testDataNested.indexData.docs,
+                    testDataNested.indexData.vectors,
+                    tmpFile.toAbsolutePath().toString(),
+                    ImmutableMap.of(INDEX_DESCRIPTION_PARAMETER, method, KNNConstants.SPACE_TYPE, spaceType.getValue()),
+                    FAISS_NAME
+                );
+                assertTrue(tmpFile.toFile().length() > 0);
+
+                long pointer = JNIService.loadIndex(
+                    tmpFile.toAbsolutePath().toString(),
+                    ImmutableMap.of(KNNConstants.SPACE_TYPE, spaceType.getValue()),
+                    FAISS_NAME
+                );
+                assertNotEquals(0, pointer);
+
+                for (float[] query : testDataNested.queries) {
+                    KNNQueryResult[] results = JNIService.queryIndex(pointer, query, k, FAISS_NAME, null, parentIds);
+                    // Verify there is no more than one result from same parent
+                    Set<Integer> parentIdSet = toParentIdSet(results, idToParentIdMap);
+                    assertEquals(results.length, parentIdSet.size());
+                }
+            }
+        }
+    }
+
+    private Set<Integer> toParentIdSet(KNNQueryResult[] results, Map<Integer, Integer> idToParentIdMap) {
+        Set<Integer> parentIdSet = new HashSet<>();
+        for (KNNQueryResult result : results) {
+            parentIdSet.add(idToParentIdMap.get(result.getId()));
+        }
+        return parentIdSet;
```

Comment:
not a strong opinion, just same can be done with one-liner similar to below:

```
return Arrays.stream(results)
   .map(result -> idToParentIdMap.get(result.getId()))
   .collect(Collectors.toSet());
```

Reply:
ack